service/history/workflow/workflow_task_state_machine.go

```diff
@@ -233,20 +233,34 @@ func (m *workflowTaskStateMachine) ReplicateWorkflowTaskTimedOutEvent(
 }
 
 func (m *workflowTaskStateMachine) AddWorkflowTaskScheduleToStartTimeoutEvent(
-	scheduledEventID int64,
+	workflowTask *WorkflowTaskInfo,
 ) (*historypb.HistoryEvent, error) {
 	opTag := tag.WorkflowActionWorkflowTaskTimedOut
-	if m.ms.executionInfo.WorkflowTaskScheduledEventId != scheduledEventID || m.ms.executionInfo.WorkflowTaskStartedEventId > 0 {
+	if m.ms.executionInfo.WorkflowTaskScheduledEventId != workflowTask.ScheduledEventID || m.ms.executionInfo.WorkflowTaskStartedEventId > 0 {
 		m.ms.logger.Warn(mutableStateInvalidHistoryActionMsg, opTag,
 			tag.WorkflowEventID(m.ms.GetNextEventID()),
 			tag.ErrorTypeInvalidHistoryAction,
-			tag.WorkflowScheduledEventID(scheduledEventID),
+			tag.WorkflowScheduledEventID(workflowTask.ScheduledEventID),
 		)
 		return nil, m.ms.createInternalServerError(opTag)
 	}
 
+	if workflowTask.Type == enumsspb.WORKFLOW_TASK_TYPE_SPECULATIVE {
+		m.ms.RemoveSpeculativeWorkflowTaskTimeoutTask()
+
+		// Create corresponding WorkflowTaskScheduled event for speculative WT.
+		scheduledEvent := m.ms.hBuilder.AddWorkflowTaskScheduledEvent(
+			m.ms.TaskQueue(),
+			workflowTask.WorkflowTaskTimeout,
+			workflowTask.Attempt,
+			timestamp.TimeValue(workflowTask.ScheduledTime).UTC(),
+		)
+		workflowTask.ScheduledEventID = scheduledEvent.GetEventId()
+		m.ms.hBuilder.FlushAndCreateNewBatch()
```

Comment:
Am I reading this correctly that speculative workflow task timeout will write the WFT events?

Reply:
yes, this is the same behavior as failed workflow tasks: first failure/timeout writes 3 events, then normal workflow task is scheduled, but because attempt counter is already 2, it automatically becomes transient.

Reply:
Can this WFT just go away here?